jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/TransitNodeRoutingShortestPathTest.java

```diff
@@ -0,0 +1,229 @@
+        assertEquals(expectedPath4, shortestPath.getPath(v1, v3));
+        assertNull(shortestPath.getPath(v3, v1));
+    }
+
+    @Test
+    public void testOnRandomGraphs() {
+        int numOfVertices = 30;
+        int vertexDegree = 5;
+        int numOfIterations = 20;
+        int source = 0;
+        Random random = new Random(SEED);
+        for (int i = 0; i < numOfIterations; ++i) {
+            testOnGraph(generateRandomGraph(numOfVertices, vertexDegree * numOfVertices, random), source);
+        }
+    }
+
+    /**
+     * Test correctness of {@link TransitNodeRoutingShortestPath} on {@code graph}
+     * starting at {@code source}.
+     *
+     * @param graph  graph
+     * @param source vertex in {@code graph}
+     */
+    private void testOnGraph(Graph<Integer, DefaultWeightedEdge> graph, Integer source) {
+        ShortestPathAlgorithm.SingleSourcePaths<Integer, DefaultWeightedEdge> dijkstraShortestPaths =
+                new DijkstraShortestPath<>(graph).getPaths(source);
+
+        ContractionHierarchy<Integer, DefaultWeightedEdge> contractionHierarchy
+                = new ContractionHierarchyPrecomputation<>(graph, () -> new Random(SEED)).computeContractionHierarchy();
+
+        TransitNodeRouting<Integer, DefaultWeightedEdge> routing =
+                new TransitNodeRoutingPrecomputation<>(contractionHierarchy).computeTransitNodeRouting();
+
+        TransitNodeRoutingShortestPath<Integer, DefaultWeightedEdge> transitNodeRoutingShortestPath
+                = new TransitNodeRoutingShortestPath<>(routing);
+        ShortestPathAlgorithm.SingleSourcePaths<Integer, DefaultWeightedEdge> tnrShortestPaths
+                = transitNodeRoutingShortestPath.getPaths(source);
+
+
+        assertEqualPaths(dijkstraShortestPaths, tnrShortestPaths, graph.vertexSet());
+    }
+
+    /**
+     * Generates an instance of random graph with {@code numOfVertices} vertices
+     * and {@code numOfEdges} edges.
+     *
+     * @param numOfVertices number of vertices
+     * @param numOfEdges    number of edges
+     * @return generated graph
+     */
+    private Graph<Integer, DefaultWeightedEdge> generateRandomGraph(int numOfVertices, int numOfEdges, Random random) {
+        DirectedWeightedPseudograph<Integer, DefaultWeightedEdge> graph =
+                new DirectedWeightedPseudograph<>(DefaultWeightedEdge.class);
+        graph.setVertexSupplier(SupplierUtil.createIntegerSupplier());
+
+        GraphGenerator<Integer, DefaultWeightedEdge, Integer> generator =
+                new GnmRandomGraphGenerator<>(numOfVertices, numOfEdges - numOfVertices + 1, SEED);
+        generator.generateGraph(graph);
+        makeConnected(graph);
+        addEdgeWeights(graph, random);
+
+        return graph;
+    }
+
+    /**
+     * Makes {@code graph} connected.
+     *
+     * @param graph graph
+     */
+    private void makeConnected(Graph<Integer, DefaultWeightedEdge> graph) {
+        Object[] vertices = graph.vertexSet().toArray();
```

Comment:
Consider using `new ArrayList<>(graph.vertexSet())` as this seems to be a bad style.